mesonbuild/ast/introspection.py

```diff
@@ -157,19 +154,21 @@ def do_subproject(self, dirname: str) -> None:
         except (mesonlib.MesonException, RuntimeError):
             return
 
-    def func_add_languages(self, node: BaseNode, args: T.List[TYPE_nvar], kwargs: T.Dict[str, TYPE_nvar]) -> None:
+    def func_add_languages(self, node: BaseNode, args: T.List[TYPE_var], kwargs: T.Dict[str, TYPE_var]) -> UnknownValue:
         kwargs = self.flatten_kwargs(kwargs)
         required = kwargs.get('required', True)
         if isinstance(required, cdata.UserFeatureOption):
             required = required.is_enabled()
+        assert isinstance(required, bool)
         if 'native' in kwargs:
             native = kwargs.get('native', False)
             self._add_languages(args, required, MachineChoice.BUILD if native else MachineChoice.HOST)
         else:
             for for_machine in [MachineChoice.BUILD, MachineChoice.HOST]:
                 self._add_languages(args, required, for_machine)
+        return UnknownValue()
```

Comment:
The explanation for `UnknownValue` was for cases where we couldn't know what would be returned. I'm not sure why I follow that `add_languages` returns `UnknownValue` and not `bool`. If it returns it will always return a `bool` (or it will abort, but for the purpose of the rewriter that doesn't really matter, does it?)

Reply:
```
var = add_languages('rust', required: false)
message(var)
```
prints `true` on some machines and `false` on others.
So `func_add_languages` has to return `UnknownValue`.

Reply:
It returns a boolean? I'm just not understanding here, if `add_languages()`, which returns a `bool` cannot be statically determined, then what can? There are no functions I know of that always return the same value in every case.

Reply:
You seem confused, let me clear that up:

You know that if you run `meson setup builddir`, the contents of the resulting `builddir`-directory depend on 

1. The contents of meson.build
2. The machine you are using.

If for example both you and I clone the same project that is using meson, and we both run `meson setup builddir`, those two directories are not (necessarily) identical if we have different machines. This is not a bug, this is intentional. Therefore, if we both run  `meson introspect builddir`, we (might) get different results.

But if we both clone the same project and run `meson introspect meson.build` we get the same result. If we don't, that is a bug. In other words, the job of `ast/introspection.py` is to know what happens if we run `meson setup` on a different, unknown machine. In other words, `ast/introspection.py` has full knowledge about the contents of `meson.build`, but no knowledge about the build/host/target machine. Let's say meson.build contains:

```
srcs = ['1.c']
srcs += files('2.c')
if 3+4 == 7
    srcs += '3.c'
endif
if build_machine.system == 'linux'
    srcs += 'linux-specific.c'
endif
if add_languages('rust', required: false)
    srcs += 'rust.rs'
endif
executable('foo', srcs)
```

If I run `meson introspect meson.build` on *my* machine, the job of meson is to figure out what sources belong to the foo-executable file if you run `meson setup` on *your* machine.
It knows that the foo-executable contains the sources 1.c and 2.c. It does not know that it contains '3.c', since I was too lazy to implement that. There is no way for it to know whether the foo-executable contains the sources 'linux-specific.c' or 'rust.rs', since a program running on *my* machine cannot know whether *your* machine has a rust compiler installed.